protokt-reflect/src/jvmMain/kotlin/protokt/v1/google/protobuf/KtMessages.kt

```diff
@@ -0,0 +1,317 @@
+import protokt.v1.Converter
+import protokt.v1.KtEnum
+import protokt.v1.KtGeneratedFileDescriptor
+import protokt.v1.KtGeneratedMessage
+import protokt.v1.KtMessage
+import protokt.v1.reflect.ClassLookup
+import protokt.v1.reflect.FieldType
+import protokt.v1.reflect.WellKnownTypes
+import protokt.v1.reflect.inferClassName
+import protokt.v1.reflect.resolvePackage
+import protokt.v1.reflect.typeName
+import kotlin.Any
+import kotlin.reflect.KClass
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.findAnnotation
+
+fun KtMessage.toDynamicMessage(context: RuntimeContext): DynamicMessage =
+    context.convertValue(this) as DynamicMessage
+
+fun KtMessage.hasField(field: FieldDescriptor): Boolean {
+    val value = getField(field)
+
+    return if (field.hasPresence()) {
+        value != null
+    } else {
+        value != defaultValue(field)
+    }
+}
+
+private fun defaultValue(field: FieldDescriptor) =
+    when (field.type) {
+        Type.UINT64, Type.FIXED64 -> 0uL
+        Type.UINT32, Type.FIXED32 -> 0u
+        Type.BYTES -> Bytes.empty()
+        else -> field.defaultValue
+    }
+
+fun KtMessage.getField(field: FieldDescriptor) =
+    ProtoktReflect.getField(this, field)
+
+class RuntimeContext internal constructor(
+    descriptors: Iterable<Descriptors.Descriptor>,
+    private val classLookup: ClassLookup
+) {
+    constructor(descriptors: Iterable<Descriptors.Descriptor>) : this(descriptors, DEFAULT_CLASS_LOOKUP)
+
+    internal val descriptorsByTypeName = descriptors.associateBy { it.fullName }
+
+    fun convertValue(value: Any?) =
+        when (value) {
+            is KtEnum -> value.value
+            is UInt -> value.toInt()
+            is ULong -> value.toLong()
+            is KtMessage -> toDynamicMessage(value, this)
+            is Bytes -> UnsafeByteOperations.unsafeWrap(value.asReadOnlyBuffer())
+
+            // pray
+            else -> value
+        }
+
+    internal fun unwrap(value: Any, field: FieldDescriptor, wrap: String): Any {
+        val proto = field.toProto()
+        val type = FieldType.from(proto.type)
+        val converterDetails =
+            classLookup.converter(
+                ClassLookup.evaluateProtobufTypeCanonicalName(
+                    proto.typeName,
+                    typeName(proto.typeName, type),
+                    type,
+                    field.name
+                ),
+                wrap
+            )
+
+        @Suppress("UNCHECKED_CAST")
+        val converter = converterDetails.converter as Converter<Any, Any>
+        return converter.unwrap(value)
+    }
+
+    companion object {
+        private val DEFAULT_CLASS_LOOKUP by lazy { ClassLookup(emptyList()) }
+
+        private val reflectiveContext by lazy {
+            RuntimeContext(getDescriptors(), DEFAULT_CLASS_LOOKUP)
+        }
+
+        fun getContextReflectively() =
+            reflectiveContext
+    }
+}
+
+private fun getDescriptors() =
+    ClassGraph()
```

Comment:
can we generate an annotation on the message class that points to the descriptor?

I feel really uncomfortable scanning the whole classpath for this

Reply:
Perhaps. We can't point to it directly, since it won't exist in lite code, but maybe there's a way to make reflection direct instead of scanned.

Reply:
Okay - the reason this is implemented reflectively is for convenience. If we had an annotation on a message in your hypothetical case, you still have to get those annotations, either by scanning or by explicit reference. If you have them explicitly then you may as well reference the descriptors themselves instead to build your own RuntimeContext.

If we don't want to encourage this kind of scan, which should only be done once, e.g. at application initialization to build validators, then we can remove this code and require users to write this themselves.